### PR TITLE
Add open api tests

### PR DESCRIPTION
**Pull request by ebh**
Adds a very basic test using [Stoplight's Prism](https://stoplight.io/open-source/prism) and the [Open API](https://swagger.io/specification/v2/) (formally known as Swagger) documentation from Bitbucket ([https://bitbucket.org/api/swagger.json](https://bitbucket.org/api/swagger.json)). Along with a Continuous Integration (CI) process using [Github Actions](https://github.com/features/actions) to continually run the tests.

A couple of things to note when reviewing:
* I was unable to use the [Services](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idservices) feature of Github Actions. This is because it does not currently support passing the `COMMAND` & `ARGS` (see [docs](https://docs.docker.com/engine/reference/commandline/create/)) parts of the `docker create` command when running a service container. I've submitted a feature request for this but in the interim, we have to start the container ourselves and pause while we wait for it to finish starting up. This has the potential to create a race condition, making the CI process unreliable. I tested with 30 runs and found a 5-second delay is sufficient.
* Because it will take time to create enough test coverage using this new testing strategy I've created the new tests to exist in parallel with the existing tests. Until such time as we have sufficient coverage and we can remove the existing tests.
* I've found bugs in the Open API documentation from Bitbucket, it does not match what they have actually implemented. So far this was just in some areas that are not critical. But I've not done an extensive comparison and there could be critical areas which block or slows us using this new strategy. Note: This is just a warning, not an "I'm sure this is not going to work so we should stop" :)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+name: Continuous Integration
```
Comment (ebh) [inline .github/workflows/ci.yml:1]: If unfamiliar with the Github Actions Workflow syntax please refer to the [docs](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions) :)

File: client.go [hunk 2]
```diff
@@ -23,6 +23,16 @@ import (
 
 const DEFAULT_PAGE_LENGTH = 10
 const DEFAULT_MAX_DEPTH = 1
+const DEFAULT_BITBUCKET_API_BASE_URL = "https://api.bitbucket.org/2.0"
```
Comment (ebh) [inline client.go:26]: Not sure if this is the best way to solve this. But it works and so thought I would put it in the PR so people with more experience than I can provide feedback :)

File: user.go [hunk 3]
```diff
@@ -21,7 +21,7 @@ type User struct {
 
 // Profile is getting the user data
 func (u *User) Profile() (*User, error) {
-	urlStr := u.c.GetApiBaseURL() + "/user/"
+	urlStr := u.c.GetApiBaseURL() + "/user"
```
Comment (ebh) [inline user.go:24]: The Open API docs from BB require that there is no trailing `/`.

**Comment by sl4mmy** [comment]
Why not update the existing tests rather than adding a new `openApiTests` dir?

In fact, because of the way you allow the default API URL to be overridden by env vars you could modify the current `test` target in `Makefile` to be something like `DEFAULT_BITBUCKET_API_BASE_URL=localhost:12345/whatever go test -v ./tests` and add a new `integration-test` target that runs the exact same tests against the real API (i.e. doesn't override the default base URL).

**Comment by sl4mmy** [comment]
Also, split the GitHub Actions into a separate PR?  These seem like separate concerns.

**Comment by ebh** [comment]
> Why not update the existing tests rather than adding a new openApiTests dir?

Great question. I didn't link to the discussions that lead to me raising this PR. In #111 & #96 there were discussions with @ktrysmt about giving this strategy a go. I'm not really wedded to any particular strategy as I'm a bit of a newbie with Go. Instead, I'm just offering my time to follow his direction so as to build up experience :) So I think your question might be better answered by him, but I would like to follow the discussion so I learn :)

> In fact, because of the way you allow the default API URL to be overridden by env vars you could modify the current test target in Makefile to be something like DEFAULT_BITBUCKET_API_BASE_URL=localhost:12345/whatever go test -v ./tests and add a new integration-test target that runs the exact same tests against the real API (i.e. doesn't override the default base URL).

What I've understood you to propose sounds very similar to my first PR, #112. Please see @ktrysmt's response to that PR. Or have I misunderstood what you are proposing?

> Also, split the GitHub Actions into a separate PR? These seem like separate concerns.

This PR is really about adding testing that can be reliability run using Github Actions. As in adding that capability is the starting point and the other changes are in support of that. That is why there is only one very simple test added. I couldn't immediately see another test I could implement that was in line with the strategy that @ktrysmt wanted and required fewer changes. But happy to stand corrected on that :)

**Comment by ktrysmt** [comment]
Thak you for your codeing and considering. @ebh @sl4mmy 

I would like to you to separate the concept into github-aciton and test codes.

At first, update test code and replacer of an api url in `./tests/`. It is more user-friendly if it is unified with `./tests/`. After checking the stability of the updated test codes, it is better to make a github action's configuration.

**Comment by ebh** [comment]
I'm going to repeat what I've understood you to say. To make sure I've understood correctly :)

You want to:
* Not introduce CI via Github Actions at this time, instead, introducing later, after a significant number of tests using Prism have been created and merged to master
* Have the existing tests, that make calls directly to the real API and the new tests that use Prism both in the same folder, `./tests/`

Is that correct?

**Comment by ktrysmt** [comment]
> Not introduce CI via Github Actions at this time, instead, introducing later, after a significant number of tests using Prism have been created and merged to master

That's right. 

> Have the existing tests, that make calls directly to the real API and the new tests that use Prism both in the same folder, ./tests/

I found a replacer of the api-base-url in your concept [here][1]. It is very make sence. I think using this will resolves testability both real bitbucket's api  and launched swagger local api server. I'd like to try to design that in Makefile;
```
$ make test
(run as)
go test -v ./...

$ make test/swagger
(run as)
DEFAULT_BITBUCKET_API_BASE_URL=http://localhost:4010 go test -v ./...
```

[1]: https://github.com/ktrysmt/go-bitbucket/pull/122/files#diff-4b667feae66c9d46b21b9ecc19e8958cf4472d162ce0a47ac3e8386af8bbd8cfR26-R138

**Comment by ebh** [comment]
Oh, so you want to migrate the existing tests so the same test runs against both Prism and the real API?

Sorry if I'm being a bit slow in understanding your intention here :)

**Comment by ktrysmt** [comment]
Yes, I like it. It is Ideally that we can switch the test target by replacing the client settings.

**Comment by ebh** [comment]
Hey @ktrysmt, sorry for the delay. School holidays here and was taking care of my nephews for a bit.

I've updated the PR as requested :)

File: Makefile [hunk 4]
```diff
@@ -17,7 +17,10 @@ endif
...
 	go test -v ./tests
 
+test/swagger:
+	env BITBUCKET_API_BASE_URL=http://0.0.0.0:4010 go test -v ./...
+
 help: ## print this help
 	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
 
-.PHONY: test help
+.PHONY: test test/swaggert help
```
Comment (ktrysmt) [inline Makefile:26]:
```suggestion
.PHONY: test test/swagger help
```

**Review by ktrysmt — COMMENTED** [review]
Thank you! 
I left a comment, please check it.

File: Makefile (hunk 4 above)
Comment (ebh) [reply Makefile]: LOL. And that is why we do peer reviews :)

**Pull request merged**